plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+		return nil, err
+	}
+
+	native, _, err := codec.NativeFromBinary(binaryData)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := p.createMetric(native, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{m}, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	metrics, err := p.Parse([]byte(line))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf("line contains multiple metrics")
+	}
+
+	return metrics[0], nil
+}
+
+func (p *Parser) SetDefaultTags(tags map[string]string) {
+	p.DefaultTags = tags
+}
+
+func (p *Parser) parseTimestamp(timestamp interface{}) (time.Time, error) {
+	if timestamp == nil {
+		return p.TimeFunc(), nil
+	}
+
+	if p.TimestampFormat == "" {
+		return p.TimeFunc(), fmt.Errorf("must specify timestamp format")
+	}
+
+	metricTime, err := internal.ParseTimestamp(p.TimestampFormat, timestamp, "UTC")
+	if err != nil {
+		return p.TimeFunc(), err
+	}
+
+	return metricTime, nil
+}
+
+func (p *Parser) createMetric(native interface{}, schema string) (telegraf.Metric, error) {
+	deep, ok := native.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("cannot cast native interface {} to map[string]interface{}")
+	}
+
+	metricTime, err := p.parseTimestamp(nestedValue(deep[p.Timestamp]))
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	for _, tag := range p.Tags {
+		if value, ok := deep[tag]; ok {
+			tags[tag] = fmt.Sprintf("%v", nestedValue(value))
+		} else {
+			p.Log.Infof("tag %s value was %v; not added to tags", tag, value)
+		}
+	}
+	fieldList := make([]string, len(p.Fields), (cap(p.Fields)+1)*2)
+	copy(fieldList, p.Fields)
+	if len(fieldList) == 0 { // Get fields from whatever we just unpacked
+		for k := range deep {
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	for _, field := range fieldList {
+		if value, ok := deep[field]; ok {
+			fields[field] = nestedValue(value)
+		} else {
+			p.Log.Infof("field %s value was %v; not added to fields", field, value)
+		}
+	}
+	var schemaObj map[string]interface{}
+
+	err = json.Unmarshal([]byte(schema), &schemaObj)
+	if err != nil {
+		return nil, err
+	}
+	if len(fields) == 0 {
+		return nil, fmt.Errorf("number of fields is 0; unable to create metric")
+	}
+	name := ""
+	if p.Measurement != "" {
+		name = p.Measurement
+	} else {
+		// get Measurement name from schema
+		nsStr, ok := schemaObj["namespace"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine namespace from schema %s", schema)
+		}
+		nStr, ok := schemaObj["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine name from schema %s", schema)
+		}
+		name = nsStr + "." + nStr
+	}
+	if name == "" {
+		name = p.MetricName
+	}
+	if name == "" {
+		return nil, fmt.Errorf("could not determine measurement name")
+	}
```

Comment:
I'm pretty sure this can be done in `Init()`.

Reply:
It can't, because the schema can (and does) change between messages.  So if we haven't explicitly set the Measurement name, we have to unpack it from the schema (as `<namespace>.<name>`) .  If somehow that fails (which it should not for a correctly-formed schema), then we use MetricName as a fallback, and if we don't have even that, then we have no idea what the metric name should be.

This logic does, however, need a bit of rework, because the namespace is allowed to be empty (https://avro.apache.org/docs/1.11.1/specification/)

Reply:
Ok I see, please also add a small comment to help the reader.

Reply:
Done.